app/server/controllers/login.ts

```diff
@@ -25,54 +45,74 @@ function getLogout(req: Request, res: Response) {
-  return async(req: Request, res: Response, next: NextFunction) => {
-    const email = req.body['email-address'];
-    const validationMessage = loginEmailAddressValidation(email);
-    if (validationMessage) {
-      const emailAddress = {
-        value: email,
-        error: validationMessage
-      };
-      return res.render('login.html', { emailAddress });
+function getIdamCallback(
+  getToken: (code: string, protocol: string, hostname: string) => TokenResponse,
+  getUserDetails: (accessToken: string) => UserDetails,
+  getOnlineHearing: (email:string) => superAgent.Response,
+  getRedirectUrl: (protocol: string, hostname: string) => string) {
+  return async (req: Request, res: Response, next: NextFunction) => {
+
+    const code: string = req.query.code;
+
+    console.log("Code: " + code);
+
+    if (!code) {
+      return getLogin(getRedirectUrl, '/login')(req, res);
     }
+
     try {
-      const response = await getOnlineHearing(email);
-      if (response.status === NOT_FOUND || response.status === UNPROCESSABLE_ENTITY) {
-        logger.info(`Know issue trying to find hearing for ${email}, status ${response.status}`);
-        const emailAddress = {
-          value: email,
-          error: i18n.login.emailAddress.error[`error${response.status}`]
-        };
-        return res.render('login.html', { emailAddress });
-      }
-      const hearing: OnlineHearing = response.body;
-      req.session.hearing = hearing;
-      logger.info(`Logging in ${email}`);
-      if (hearing.decision && hearing.decision.decision_state === CONST.DECISION_ISSUED_STATE) {
-        logger.info(`Decision issued for hearing with ID ${hearing.online_hearing_id}`);
-        return res.redirect(Paths.decision);
-      }
-      return res.redirect(Paths.taskList);
+      const tokenResponse: TokenResponse = await getToken(code, req.protocol, req.hostname);
+      const userDetails: UserDetails = await getUserDetails(tokenResponse.access_token);
+      // todo Maybe need to check userDetails.accountStatus is 'active' and userDetails.roles contains 'citizen' on userDetails
+
+      const email: string = userDetails.email;
+
+      return await loadHearingAndEnterService(getOnlineHearing, email, req, res)
     } catch (error) {
       AppInsights.trackException(error);
       return next(error);
     }
   };
 }
 
+async function loadHearingAndEnterService(
+  getOnlineHearing: (email:string) => superAgent.Response,
+  email: string,
+  req: Request,
+  res: Response) {
+  const response: superAgent.Response = await getOnlineHearing(email);
+  if (response.status === NOT_FOUND || response.status === UNPROCESSABLE_ENTITY) {
+    logger.info(`Know issue trying to find hearing for ${email}, status ${response.status}`);
+    const emailAddress: any = {
+      value: email,
+      error: i18n.login.emailAddress.error[`error${response.status}`]
+    };
+    return res.render('login.html', {emailAddress}); //todo this is incorrect
```

Comment:
We need to add a new story to capture handling these errors and showing them to the user in some way.

Reply:
I've updated so it doesn't go to the login page as this will let you in by the back door. Have spoken with UI guys and a few issues came up. We will run a spike to look at them need to create the Jira for that and then this can come out of that.

Reply:
Sounds good to me